File: backend/src/routes/oa_announcement.rs
//! OA 公告路由（P0-D17 / Batch 488）
//!
//! 7 端点（5 基础 CRUD + 2 状态转换）：
//! - GET    /                  列表（支持 status/announcement_type/is_top 过滤）
//! - POST   /                  创建（默认 DRAFT 状态）
//! - GET    /:id               详情
//! - PUT    /:id               更新（DRAFT 全字段，PUBLISHED/ARCHIVED 仅 expiry_date/remarks/is_top）
//! - DELETE /:id               删除（仅 DRAFT 可硬删除）
//! - POST   /:id/publish       发布（DRAFT → PUBLISHED）
//! - POST   /:id/archive       归档（PUBLISHED → ARCHIVED）
//!
//! 路由注册顺序：静态路径（publish/archive）在 /:id 之前会冲突，
//! axum 路由树允许 /:id 与 /:id/publish 共存（前者是单段，后者是两段），
//! 因此无须特殊顺序。

use axum::{
    routing::{delete, get, post, put},
    Router,
};

use crate::handlers::oa_announcement_handler;
use crate::utils::app_state::AppState;

/// OA 公告路由（nest 到 /api/v1/erp/oa-announcements）
pub fn routes() -> Router<AppState> {
    Router::new()
        .route("/", get(oa_announcement_handler::list))
        .route("/", post(oa_announcement_handler::create))
        .route("/:id", get(oa_announcement_handler::get))
        .route("/:id", put(oa_announcement_handler::update))
        .route("/:id", delete(oa_announcement_handler::delete))
        .route("/:id/publish", post(oa_announcement_handler::publish))
        .route("/:id/archive", post(oa_announcement_handler::archive))
}

File: frontend/src/api/oa-announcement.ts
import { request } from './request';
import type { ApiResponse } from '@/types/api';

export interface OaAnnouncement {
  id?: number;
  title: string;
  content: string;
  announcement_type: string;
  publish_date: string;
  effective_date: string;
  expiry_date?: string | null;
  publisher_id?: number;
  status: string;
  is_top: boolean;
  attachments?: unknown;
  remarks?: string | null;
  visibility_scope: string;
  visible_scope_config?: {
    user_ids?: number[];
    department_ids?: number[];
    role_ids?: number[];
  } | null;
  created_at?: string;
  updated_at?: string;
  notified_count?: number;
}

export interface CreateOaAnnouncementRequest {
  title: string;
  content: string;
  announcement_type: string;
  publish_date: string;
  effective_date: string;
  expiry_date?: string;
  remarks?: string;
  is_top?: boolean;
  visibility_scope: string;
  visible_scope_config?: unknown;
}

export interface UpdateOaAnnouncementRequest {
  title?: string;
  content?: string;
  announcement_type?: string;
  expiry_date?: string;
  remarks?: string;
  is_top?: boolean;
  visibility_scope?: string;
  visible_scope_config?: unknown;
}

// 公告列表查询参数：字段集严格对齐后端 OaAnnouncementQuery（services/oa_announcement_service.rs）。
// 后端不读 keyword/order_by/order_dir/date_range 等通用键。
export interface OaAnnouncementQuery {
  /** DRAFT/PUBLISHED/ARCHIVED */
  status?: string;
  /** NOTICE/ANNOUNCEMENT/NEWS */
  announcement_type?: string;
  is_top?: boolean;
  page?: number;
  page_size?: number;
}

export function listOaAnnouncements(
  params?: OaAnnouncementQuery
): Promise<ApiResponse<{ items: OaAnnouncement[]; total: number }>> {
  return request.get('/oa-announcements', { params });
}

export function getOaAnnouncement(id: number): Promise<ApiResponse<OaAnnouncement>> {
  return request.get(`/oa-announcements/${id}`);
}

export function createOaAnnouncement(
  data: CreateOaAnnouncementRequest
): Promise<ApiResponse<OaAnnouncement>> {
  return request.post('/oa-announcements', data);
}

export function updateOaAnnouncement(
  id: number,
  data: UpdateOaAnnouncementRequest
): Promise<ApiResponse<OaAnnouncement>> {
  return request.put(`/oa-announcements/${id}`, data);
}

export function deleteOaAnnouncement(id: number): Promise<ApiResponse<void>> {
  return request.delete(`/oa-announcements/${id}`);
}

export function publishOaAnnouncement(id: number): Promise<ApiResponse<OaAnnouncement>> {
  return request.post(`/oa-announcements/${id}/publish`);
}

export function archiveOaAnnouncement(id: number): Promise<ApiResponse<OaAnnouncement>> {
  return request.post(`/oa-announcements/${id}/archive`);
}
